### PR TITLE
ARROW-6165: [Integration] Run integration tests on multiple cores

### PR DESCRIPTION
**Pull request by pitrou**
On a local test with C++, Java, Flight and Gold files enabled, the runtime goes down from 1m56 (serial) to 35s (parallel).

### CODE REVIEW

**Comment by pitrou** [comment]
This PR seems to save 7 minutes on the Github Actions Integration job:
* before: https://github.com/apache/arrow/runs/443660012?check_suite_focus=true
* after: https://github.com/apache/arrow/pull/6416/checks?check_run_id=443983974

**Comment by pitrou** [comment]
@wesm @kszucs

File: dev/archery/archery/integration/util.py [hunk 1]
```diff
@@ -40,6 +45,61 @@ def guid():
...
+        assert not self._tls.corked, "reentrant call"
+        inner_stdout = self._tls.stdout = io.StringIO()
+        self._tls.corked = True
+        try:
+            yield
+        finally:
+            self._tls.stdout = outer_stdout
+            self._tls.corked = False
+            outer_stdout.write(inner_stdout.getvalue())
+            outer_stdout.flush()
```
Comment (wesm) [inline dev/archery/archery/integration/util.py]: This is nice

**Review by wesm — APPROVED** [review]
+1, this looks great, I'm happy to see this code "loved on" / cleaned up by more people. The time savings are very nice

**Pull request closed without merging**